### PR TITLE
Implement Erlang ctags.

### PR DESCRIPTION
**Pull request by bengtan**
Hi,

I've implemented ctags for Erlang for Geany. It's not perfect but it works for me and it's much better than having nothing at all. Please have a look and let me know what you think. 

Only tested on Ubuntu Linux. Not tested on other platforms. 

Critiques welcome.

Thank you.

### CODE REVIEW

**Comment by elextr** [comment]
The parsers are one of the few things that have tests, it would be good if you can add tests for this, see tests/ctags/*

**Comment by b4n** [comment]
Upstream CTags has an Erlang parser (apparently since 2003), have you looked at it?  You can check out the [Fishman fork](https://github.com/fishman/ctags) that is currently trying to resurrect CTags, and has some additions.
We'd love to keep as much in sync with CTags, so if your versions has improvements we'd love to see them contributed back there :)

Apart that, while I didn't review the parser code itself yet, the infrastructure around it is perfect :)

However apparently your changes don't merge cleanly, probably because you used an older revision of the repository and changes have happened near yours since then.  Could you rebase your changes on top of current master and fix the conflicts?  If it sounds too scary to you don't worry, we can do it too when merging.

**Comment by bengtan** [comment]
> The parsers are one of the few things that have tests, it would be 
> good if you can add tests for this, see tests/ctags/*

Noted. I'll look into it at a later time and get back.

> Upstream CTags has an Erlang parser (apparently since 2003), have you 
> looked at it? 

Uh, no, I haven't ...

 ... This is one thing I don't understand about the Geany code. 

I know there is already an Erlang parser in Geany because syntax highlighting works for Erlang. However, that seems to live somewhere else in the code and the ctags code I've written lives in another place. 

It's a bit confusing why there are two "parser systems" in Geany.

*Snip* See below

> Apart that, while I didn't review the parser code itself yet, 
> the infrastructure around it looks perfect :)

For the Windows makefiles, I was just going on the documented instructions so it's good to know they're correct. 

> However apparently your changes don't merge cleanly, probably because
> you used an older revision of the repository and changes have happened 
> near yours since then. 

Yes, because I was working off the last stable release 1.24.1. I'll come back to this.

> You can check out the Fishman fork that is currently trying to
> resurrect CTags, and has some additions.

That's very interesting. I instantly recognise it as being very similar to what I've done. His is probably better :) 

I'll look into it and come back to this.

**Comment by bengtan** [comment]
> The parsers are one of the few things that have tests, it would be 
> good if you can add tests for this, see tests/ctags/*

Noted. I'll look into it at a later time and get back.

> Upstream CTags has an Erlang parser (apparently since 2003), have you 
> looked at it? 

Uh, no, I haven't ...

 ... This is one thing I don't understand about the Geany code. 

I know there is already an Erlang parser in Geany because syntax highlighting works for Erlang. However, that seems to live somewhere else in the code and the ctags code I've written lives in another place. 

It's a bit confusing why there are two "parser systems" in Geany.

*Snip* See below

> Apart that, while I didn't review the parser code itself yet, 
> the infrastructure around it looks perfect :)

For the Windows makefiles, I was just going on the documented instructions so it's good to know they're correct. 

> However apparently your changes don't merge cleanly, probably because
> you used an older revision of the repository and changes have happened 
> near yours since then. 

Yes, because I was working off the last stable release 1.24.1. I'll come back to this.

> You can check out the Fishman fork that is currently trying to
> resurrect CTags, and has some additions.

That's very interesting. I instantly recognise it as being very similar to what I've done. His is probably better :) 

I'll look into it and come back to this.

**Comment by elextr** [comment]
On 9 March 2015 at 17:53, Beng Tan <notifications@github.com> wrote:

> The parsers are one of the few things that have tests, it would be
> good if you can add tests for this, see tests/ctags/*
>
> Noted. I'll look into it at a later time and get back.
>
> Upstream CTags has an Erlang parser (apparently since 2003), have you
> looked at it?
>
> Uh, no, I haven't ...
>
> ... This is one thing I don't understand about the Geany code.
>
> I know there is already an Erlang parser in Geany because syntax
> highlighting works for Erlang. However, that seems to live somewhere else
> in the code and the ctags code I've written lives in another place.
>
> It's a bit confusing why there are two "parser systems" in Geany.
>

​Hi,

​The highlighting is syntax based, so it uses only a lexer, not a full
parser to make it fast as it has to change each keystroke.  This lexer is
part of the Scintilla editing component we use (see www.scintilla.org) and
is in the Scintilla directory.

But some of the highlighting of identifiers can be modified by semantic
information provided by Geany which it gets from the parsers which identify
semantic information such as which identifiers represent types and other
symbol semantics.  The parsers are based on the ctags ones and exist in the
tagmanager directory.  They can be slower since they do not run every
keystroke.

Cheers
Lex

**Comment by bengtan** [comment]
> The highlighting is syntax based, so it uses only a lexer

> The parsers are based on the ctags ones and exist in the tagmanager 
> directory. They can be slower since they do not run every keystroke.

Thanks for the clarification. I guess that makes sense. 

I've had a look at erlang.c in the fishman ctags repository. I didn't know it existed, otherwise I would have used that instead of creating something from scratch. As it is, my code was just a quick hack so I'd prefer to make some additions to his instead. 

What do you suggest is the proper process for this? 

A. Make some additions to erlang.c. Get it committed to geany. Notify the upstream repository (ie. fishman ctags) and ask for a merge?

or 

B. Make some additions to erlang.c. Notify the upstream repository (ie. fishman ctags) and ask for a merge. Then backport it to geany?

I'm in favour of (A) because I don't really want to have to build yet another repository that I know nothing about.

**Comment by elextr** [comment]
On 12 March 2015 at 17:41, Beng Tan <notifications@github.com> wrote:

> The highlighting is syntax based, so it uses only a lexer
>
> The parsers are based on the ctags ones and exist in the tagmanager
> directory. They can be slower since they do not run every keystroke.
>
> Thanks for the clarification. I guess that makes sense.
>
> I've had a look at erlang.c in the fishman ctags repository. I didn't know
> it existed, otherwise I would have used that instead of creating something
> from scratch. As it is, my code was just a quick hack so I'd prefer to make
> some additions to his instead.
>
> What do you suggest is the proper process for this?
>
> A. Make some additions to erlang.c. Get it committed to geany. Notify the
> upstream repository (ie. fishman ctags) and ask for a merge?
>
> or
>
> B. Make some additions to erlang.c. Notify the upstream repository (ie.
> fishman ctags) and ask for a merge. Then backport it to geany?
>
> I'm in favour of (A) because I don't really want to have to build yet
> another repository that I know nothing about.
>

​Well, its unlikely that anyone else is going to push it upstream, or into
Geany, so its up to you.  I guess more people will use it if its in
upstream, allowing you to also benefit from their
improvements/corrections.  So whichever way is easiest for you.

Cheers
Lex
​

**Comment by b4n** [comment]
> […] What do you suggest is the proper process for this?

I prefer solution B (change ctags, then import new version), especially because they already have some tests for that parser (`make check LANGUAGES=Erlang` for fishman's), so it's easy to check whether it don't break anything (seems easier to me than importing the checks in Geany to then check).

> I'm in favour of (A) because I don't really want to have to build yet another repository that I know nothing about. 

If it's too much hassle for you, you can do A, sure.  But building fishman-ctags is very easy, and you don't have to know much more than what you already learned writing the parser for Geany :)
Only important additional thing is that checks are located in *Units*, documented in *doc/units.rst* and can be run automatically with `make check LANGUAGES=LangA,LangB,LangC`.

Basically, after you cloned the repository, just do:
```console
$ autoreconf -vfi  # documented in autogen.sh
$ ./configure  # with options if you want
$ make check LANGUAGES=Erlang
```

And then, alter *erlang.c* as you want, and re-run `make check LANGUAGES=Erlang` to test it.  You of course can/should add tests for things you fix/add, but that's basically just writing a source file and checking the generated tags are correct.

**Comment by b4n** [comment]
@bengtan BTW, I'm on fishman-ctags, so don't worry about a need to get "them" to review the patch and merge it, I'll do it ;)

**Comment by bengtan** [comment]
I found it easier to create a new branch.

Please continue this discussion at https://github.com/geany/geany/pull/445. Thanks.
